### PR TITLE
Backend support for configurable record summaries

### PR DESCRIPTION
**Pull request by seancolsen**
## Notes

1. I implemented DB-layer functionality to render record summaries based on custom templates passed into the following high-level DB functions:

    - list_records_from_table
    - search_records_from_table
    - get_record_from_table
    - add_record_to_table
    - patch_record_in_table

1. I implemented storage of table metadata to persist record summary templates in the service layer by associating a JSON template with a table OID.

1. I exposed _one_ new param at the API layer — a `table_record_summary_templates` param on `records.get` which allows the client to render one record summary using an ephemeral template.

1. I added some small SQL tests to cover new record summary template logic.

1. I didn't bother with any tests in the service layer.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by seancolsen** [comment]
This state due to some rough edges and unresolved architectural questions, but I've got some progress...

image

In this `records.list` request, I'm passing in _all_ the record summary templates to the DB layer. This was one possible implementation that we discussed, but not necessarily one that I feel strongly about using. It just happened to be the easiest starting place.

File: hardcopy.0
Comment (mathemancer) [inline hardcopy.0:1]: I think this file should be removed.

**Review by mathemancer — CHANGES_REQUESTED** [review]
The only blocking change is to remove the file `hardcopy.0`, unless I'm missing something.

Other than that, I think I'm good with this.

File: hardcopy.0
Comment (seancolsen) [reply hardcopy.0:1]: Oh geeze, how embarrassing. 🤣 That's a dump from GNU Screen that probably got there from me fat-fingering a keyboard shortcut. Someday I'll finally switch to tmux 🙄 I've removed the file in 2397661a89ff7cd9bb02e65d4669043baf6b4051

File: db/sql/00_msar.sql [hunk 1]
```diff
@@ -4809,8 +5040,8 @@ SELECT 'jsonb_build_object(' || string_agg(
     $j$
     %1$L, COALESCE(
       jsonb_object_agg(
-        summary_cte_%1$s.fkey, summary_cte_%1$s.summary
-      ) FILTER (WHERE summary_cte_%1$s.fkey IS NOT NULL), '{}'::jsonb
+        summary_cte_%1$s.key, summary_cte_%1$s.summary
+      ) FILTER (WHERE summary_cte_%1$s.key IS NOT NULL), '{}'::jsonb
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply db/sql/00_msar.sql:5044]: Yeah, I had to do this in order to make the _self_ summary query use the same structure as the _linked_ summary query.

File: db/sql/00_msar.sql [hunk 2]
```diff
@@ -4715,59 +4767,238 @@ LIMIT 1;
...
+  IF cardinality(expr_parts) = 0 THEN
+    -- If the template didn't give us anything to render, then we show '?' as a fallback. This can
+    -- happen if (e.g.) the template only contains a reference which is no longer valid due to a
+    -- column being deleted.
+    expr_parts := array_append(expr_parts, quote_literal('?'));
+  END IF;
+
+  join_section := CASE
+    WHEN array_length(join_clauses, 1) = 0 THEN ''
+    ELSE chr(10) || array_to_string(join_clauses, chr(10))
```
[the remark this answers is not in the document]
Comment (seancolsen) [reply db/sql/00_msar.sql]: Yeah, that's fair. I changed these to `E'\n'` in ab8161b814e2f6cb7e95baa01e87b973d483c51c

**Comment by seancolsen** [comment]
@mathemancer this is ready for re-review now.

**Review by mathemancer — APPROVED** [review]
Okay, LGTM

**Pull request merged**